Source/NSubstitute/Routing/Handlers/ReturnAutoValue.cs

```diff
@@ -13,37 +13,122 @@ public enum AutoValueBehaviour
     }
     public class ReturnAutoValue : ICallHandler
     {
-        private readonly IEnumerable<IAutoValueProvider> _autoValueProviders;
-        private readonly AutoValueBehaviour _autoValueBehaviour;
+        private readonly IAutoValueProvider _autoValueProvider;
+        private readonly Func<AutoValueBehaviour> _autoValueBehaviour;
         private readonly IConfigureCall ConfigureCall;
 
-        public ReturnAutoValue(AutoValueBehaviour autoValueBehaviour, IEnumerable<IAutoValueProvider> autoValueProviders, IConfigureCall configureCall)
+        public ReturnAutoValue(Func<AutoValueBehaviour> autoValueBehaviour, IAutoValueProvider autoValueProvider, IConfigureCall configureCall)
         {
-            _autoValueProviders = autoValueProviders;
+            _autoValueProvider = autoValueProvider;
             ConfigureCall = configureCall;
             _autoValueBehaviour = autoValueBehaviour;
         }
 
         public RouteAction Handle(ICall call)
         {
             var type = call.GetReturnType();
-            var compatibleProviders = _autoValueProviders.Where(x => x.CanProvideValueFor(type)).FirstOrNothing();
+            var compatibleProviders = _autoValueProvider.GetValue(type);
             return compatibleProviders.Fold(
-                RouteAction.Continue,
-                ReturnValueUsingProvider(call, type));
+                () => NoReturnValue(call),
+                x => ReturnValue(call, x));
         }
 
-        private Func<IAutoValueProvider, RouteAction> ReturnValueUsingProvider(ICall call, Type type)
+        private IEnumerable<ByRefArgument> GetByRefValues(CallInfo args)
         {
-            return provider =>
+            var parameters =
+                args.ArgTypes()
+                    .Select((x, i) => new { Parameter = x, Index = i })
+                    .Where(x => x.Parameter.IsByRef);
+
+            foreach (var p in parameters)
+            {
+                var type = p.Parameter;
+                var value = _autoValueProvider.GetValue(type);
+
+                if (value.HasValue())
+                    yield return new ByRefArgument(
+                        p.Index,
+                        new Lazy<object>(() => value.ValueOrDefault()));
+            }
+        }
+
+        private void SetByRefValues(CallInfo callInfo, IEnumerable<ByRefArgument> values)
+        {
+            foreach (var value in values)
+                callInfo[value.Index] = value.Value.Value;
+        }
+
+        private ReturnValueFromFunc<T> GetReturnValue<T>(T value, IEnumerable<ByRefArgument> byRefValues)
+        {
+            return new ReturnValueFromFunc<T>(callInfo =>
+            {
+                SetByRefValues(callInfo, byRefValues);
+                return value;
+            });
+        }
+
+        private RouteAction NoReturnValue(ICall call)
+        {
+            var callInfo = new CallInfoFactory().Create(call);
```

Comment:
You're right. You can't get the `ref` arguments though. That's why I had to go this way.